compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1593,45 +1599,124 @@ private BObjectType getDistinctObjectType(BLangTypeDefinition typeDefinition, BO
         return definedObjType;
     }
 
-    private BType getCyclicDefinedType(BLangTypeDefinition typeDef, SymbolEnv env) {
-        BUnionType unionType = BUnionType.create(null, new LinkedHashSet<>());
-        unionType.isCyclic = true;
-        Name typeDefName = names.fromIdNode(typeDef.name);
-
-        BTypeSymbol typeDefSymbol = Symbols.createTypeSymbol(SymTag.UNION_TYPE, Flags.asMask(typeDef.flagSet),
-                typeDefName, env.enclPkg.symbol.pkgID, unionType, env.scope.owner,
-                typeDef.name.pos, SOURCE);
-
-        typeDef.symbol = typeDefSymbol;
-        unionType.tsymbol = typeDefSymbol;
-
-        // We define the unionType in the main scope here
+    private void defineTypeInMainScope(BTypeSymbol typeDefSymbol, BLangTypeDefinition typeDef) {
         if (PackageID.isLangLibPackageID(this.env.enclPkg.packageID)) {
             typeDefSymbol.origin = BUILTIN;
             handleLangLibTypes(typeDef);
         } else {
             defineSymbol(typeDef.name.pos, typeDefSymbol);
         }
+    }
+
+    private BType defineSymbolsForCyclicTypeDefinitions(BLangTypeDefinition typeDef, SymbolEnv env) {
+        Name newTypeDefName = names.fromIdNode(typeDef.name);
+        BTypeSymbol typeDefSymbol;
+        BType newTypeNode;
+
+        switch (typeDef.typeNode.getKind()) {
+            case TUPLE_TYPE_NODE:
+                newTypeNode = new BTupleType(null, new ArrayList<>());
+                typeDefSymbol = Symbols.createTypeSymbol(SymTag.TUPLE_TYPE, Flags.asMask(typeDef.flagSet),
+                        newTypeDefName, env.enclPkg.symbol.pkgID, newTypeNode, env.scope.owner,
+                        typeDef.name.pos, SOURCE);
+                break;
+            default:
+                newTypeNode = BUnionType.create(null, new LinkedHashSet<>());
+                typeDefSymbol = Symbols.createTypeSymbol(SymTag.UNION_TYPE, Flags.asMask(typeDef.flagSet),
+                        newTypeDefName, env.enclPkg.symbol.pkgID, newTypeNode, env.scope.owner,
+                        typeDef.name.pos, SOURCE);
+        }
+        newTypeNode.isCyclic = true;
+        BSymbol foundSym = symResolver.lookupSymbolInMainSpace(env, names.fromIdNode(typeDef.name));
+        if (foundSym == symTable.notFoundSymbol) {
+            typeDef.symbol = typeDefSymbol;
+            defineTypeInMainScope(typeDefSymbol, typeDef);
+            newTypeNode.tsymbol = typeDefSymbol;
+            newTypeNode.flags |= typeDefSymbol.flags;
+            return newTypeNode;
+        }
+        newTypeNode = foundSym.type;
+        newTypeNode.tsymbol = (BTypeSymbol) foundSym;
+        newTypeNode.flags |= foundSym.flags;
+        return newTypeNode;
+    }
+
+    private void iterateMembersForCyclicTypes(List<BLangType> memberTypeNodes) {
+        for (BLangType member : memberTypeNodes) {
```

Comment:
I think we can improve this code. Worse case for each member for every cyclic union we may iterate through all the cyclic type definitions. Logic now is to always define a symbol for cyclic type definition beforehand.  Instead of iterating through the unresolved list for each member can't we create a map with names against a cyclic type definition or just a list of cyclic type definitions. We can collect this information when we flag `BLangTypeDefinition.hasCyclicReference` boolean. Then we can go through this cyclic list and define them one by one without going through member lists.

Reply:
Instead of using a map, as suggested, I first defined all unresolved the cyclic type definitions in the scope and then resolved each node.